### PR TITLE
[TableGen] Add a backend to generate MacroFusion predicators

### PR DESCRIPTION
**Pull request by wangpc-pp**
`FusionPredicate` is used to predicate if target instruction matches
 the requirement. The targets can be firstMI, secondMI or both.

The `Fusion` contains a list of `FusionPredicate`. The generated code
will be like:
```
bool isNAME(const TargetInstrInfo &TII,
            const TargetSubtargetInfo &STI,
            const MachineInstr *FirstMI,
            const MachineInstr &SecondMI) {
  auto &MRI = SecondMI.getMF()->getRegInfo();
  /* Predicates */
  return true;
}
```

A boilerplate class called `SimpleFusion` is added. `SimpleFusion` has
a predefined structure of predicates and accepts predicate for
`firstMI`, predicate for `secondMI` and epilog/prolog as arguments.
The generated code for `SimpleFusion` will be like:
```
bool isNAME(const TargetInstrInfo &TII,
            const TargetSubtargetInfo &STI,
            const MachineInstr *FirstMI,
            const MachineInstr &SecondMI) {
  auto &MRI = SecondMI.getMF()->getRegInfo();
  /* Prolog */
  /* Predicate for `SecondMI` */
  /* Wildcard */
  /* Predicate for `FirstMI` */
  /* Check One Use */
  /* Tie registers */
  /* Epilog */
  return true;
}
```

### CODE REVIEW

File: llvm/utils/TableGen/MacroFusionPredicatorEmitter.cpp [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+    OS.indent(5) << "const MachineInstr &SecondMI) {\n";
+    OS.indent(2) << "auto &MRI = SecondMI.getMF()->getRegInfo();\n";
+
+    emitFirstPredicate(First, PE, OS);
+    emitSecondPredicate(Second, PE, OS);
+
+    if (!Prolog.empty())
+      emitPrologAndEpilog(Prolog, PE, OS);
+
+    OS.indent(2) << "if(!matchFirst(FirstMI))\n";
```
Comment (topperc) [inline llvm/utils/TableGen/MacroFusionPredicatorEmitter.cpp:89]: Space after if

File: llvm/utils/TableGen/MacroFusionPredicatorEmitter.cpp [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+  OS << "\n#endif\n";
+  OS << "#undef GET_" << Target.getName() << "_MACRO_FUSION_PRED_IMPL\n\n";
+}
+
+void MacroFusionPredicatorEmitter::emitFirstPredicate(Record *FirstPredicate,
+                                                      PredicateExpander &PE,
+                                                      raw_ostream &OS) {
+
+  OS.indent(2) << "auto matchFirst = [&](const MachineInstr* MI) {\n";
```
Comment (topperc) [inline llvm/utils/TableGen/MacroFusionPredicatorEmitter.cpp:109]: `const MachineInstr *MI`

File: llvm/utils/TableGen/MacroFusionPredicatorEmitter.cpp [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+  OS.indent(4) << "return ";
+  PE.expandPredicate(OS, FirstPredicate);
+  OS << ";\n";
+  OS.indent(2) << "};\n";
+}
+
+void MacroFusionPredicatorEmitter::emitSecondPredicate(Record *SecondPredicate,
+                                                       PredicateExpander &PE,
+                                                       raw_ostream &OS) {
+  OS.indent(2) << "auto matchSecond = [&](const MachineInstr* MI) {\n";
```
Comment (topperc) [inline llvm/utils/TableGen/MacroFusionPredicatorEmitter.cpp:119]: `const MachineInstr *MI`

**Review by Sisyph — COMMENTED** [review]
I like the approach. It seems quite powerful and useful.

File: llvm/utils/TableGen/MacroFusionPredicatorEmitter.cpp [hunk 4]
```diff
@@ -0,0 +1,199 @@
+//===--- MacroFusionPredicatorEmitter.cpp - Generator for Fusion ----===//
```
Comment (mshockwave) [inline llvm/utils/TableGen/MacroFusionPredicatorEmitter.cpp:1]: nit: please match the line width with line 10

File: llvm/utils/TableGen/MacroFusionPredicatorEmitter.cpp [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+    PE.setNegatePredicate(true);
+    PE.setIndentLevel(3);
+    PE.expandPredicate(OS, Predicate->getValueAsDef("Predicate"));
+    OS << ")\n";
+    OS.indent(4) << "  return false;\n";
+    OS.indent(2) << "}\n";
+  } else
+    PrintFatalError(Predicate->getLoc(),
+                    "Unsupported predicate for first instruction: " +
+                        Predicate->getType()->getAsString());
```
Comment (mshockwave) [inline llvm/utils/TableGen/MacroFusionPredicatorEmitter.cpp:135]: format: please add braces here to be consistent with other branches

File: llvm/utils/TableGen/MacroFusionPredicatorEmitter.cpp [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+    OS.indent(2) << "{\n";
+    OS.indent(4) << "const MachineInstr *MI = &SecondMI;\n";
+    OS.indent(4) << "if (";
+    PE.setNegatePredicate(true);
+    PE.setIndentLevel(3);
+    PE.expandPredicate(OS, Predicate->getValueAsDef("Predicate"));
+    OS << ")\n";
+    OS.indent(4) << "  return false;\n";
+    OS.indent(2) << "}\n";
+  } else
```
Comment (mshockwave) [inline llvm/utils/TableGen/MacroFusionPredicatorEmitter.cpp:151]: ditto braces

**Review by mshockwave — COMMENTED** [review]
Could you also add TableGen tests?

File: llvm/utils/TableGen/MacroFusionPredicatorEmitter.cpp [hunk 7]
```diff
@@ -0,0 +1,199 @@
+//===--- MacroFusionPredicatorEmitter.cpp - Generator for Fusion ----===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
```
Comment (Pierre-vh) [inline llvm/utils/TableGen/MacroFusionPredicatorEmitter.cpp:6]: nit: the header usually has 
```
//===----------------------------------------------------------------------===//
````
Between the license and the documentation

File: llvm/utils/TableGen/MacroFusionPredicatorEmitter.cpp [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+// MacroFusionPredicatorEmitter implements a TableGen-driven predicators
+// generator for macro-op fusions.
```
Comment (Pierre-vh) [inline llvm/utils/TableGen/MacroFusionPredicatorEmitter.cpp:8]: Maybe not the best place for this but can you give a quick explanation of where/how this is used? It's not immediately obvious, and if someone has to fix a bug here it'd be easier if there was an example of what this does.

The file overall is lacking docs, none of the methods/classes have any info about them. Just having a sentence or two about the intention behind them (unless it's obvious, e.g. `MacroFusionPredicatorEmitter`)  would be a nice thing.

File: llvm/utils/TableGen/MacroFusionPredicatorEmitter.cpp [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+  // Emit file header.
+  emitSourceFileHeader("Macro Fusion Predicators", OS);
+
+  PredicateExpander PE(Target.getName());
+  PE.setByRef(false);
+  PE.setExpandForMC(false);
+
+  std::vector<Record *> Fusions = Records.getAllDerivedDefinitions("Fusion");
+  // Sort macro fusions by name.
+  llvm::sort(Fusions, LessRecord());
```
Comment (Pierre-vh) [inline llvm/utils/TableGen/MacroFusionPredicatorEmitter.cpp:193]: don't need `llvm::`

**Review by Pierre-vh — COMMENTED** [review]
Not part of the review but adding some nits/comments.

I also think it would be nice if this backend had at least one test to track the expected codegen, and see codegen changes. It makes future maintenance a lot easier if we can see what changes do to codegen during a review.

**Comment by wangpc-pp** [comment]
> Could you also add TableGen tests?

I have added a test, thanks!

**Comment by wangpc-pp** [comment]
Ping.

File: llvm/include/llvm/Target/TargetSchedule.td [hunk 10]
```diff
@@ -584,3 +584,116 @@ class MemoryQueue<ProcResourceKind PR> {
 
 class LoadQueue<ProcResourceKind LDQueue> : MemoryQueue<LDQueue>;
 class StoreQueue<ProcResourceKind STQueue> : MemoryQueue<STQueue>;
+
+// The target instruction that FusionPredicate will be evaluated on.
+class FusionTarget;
+def first : FusionTarget;
+def second : FusionTarget;
+def both : FusionTarget;
```
Comment (arsenm) [inline llvm/include/llvm/Target/TargetSchedule.td:592]: These are really general names to introduce into the global namespace; first_fusion_target maybe?

**Review by arsenm — APPROVED** [review]

**Comment by wangpc-pp** [comment]
I will land this at this Friday if there is no more comment. :-)

File: llvm/utils/TableGen/MacroFusionPredicatorEmitter.cpp [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+  OS << "#ifdef GET_" << Target.getName() << "_MACRO_FUSION_PRED_IMPL\n";
+  OS << "#undef GET_" << Target.getName() << "_MACRO_FUSION_PRED_IMPL\n\n";
+  OS << "namespace llvm {\n";
+
+  for (Record *Fusion : Fusions) {
+    std::vector<Record *> Predicates =
+        Fusion->getValueAsListOfDefs("Predicates");
+
+    OS << "bool is" << Fusion->getName() << "(\n";
+    OS.indent(5) << "const TargetInstrInfo &TII,\n";
```
Comment (topperc) [inline llvm/utils/TableGen/MacroFusionPredicatorEmitter.cpp:108]: I think 4 is the correct indent according to clang-format.

**Review by arsenm — APPROVED** [review]

**Pull request merged**